bigquery/storage/managedwriter/appendresult.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedwriter
+
+import (
+	"context"
+
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1beta2"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// AppendResult tracks the status of a single row of data.
+type AppendResult struct {
+	// rowData contains the serialized row data.
+	rowData []byte
+
+	ready chan struct{}
+
+	// if the encapsulating append failed, this will retain a reference to the error.
+	err error
+
+	// the stream offset
+	offset int64
+}
+
+func newAppendResult(data []byte) *AppendResult {
+	return &AppendResult{
+		ready:   make(chan struct{}),
+		rowData: data,
+	}
+}
+
+// Ready blocks until the append request is completed.
+func (ar *AppendResult) Ready() <-chan struct{} { return ar.ready }
+
+// GetResult returns the optional offset of this row, or the associated
```

Comment:
When this says optional offset does that mean sometimes the offset will be 0 even if there is not an error?

Reply:
There's certain kinds of streams (default, committed) where sending offset with an append isn't supported.  Additionally, for (default) streams the backend responses never tell you the current offset because the stream always exists, can can be written to by multiple streams simultaneously.  In this case, you only care about error presence.

My intent was to return 0, but perhaps the exported constant is better.